std-bits/base/src/main/java/org/enso/base/encoding/ReportingStreamDecoder.java

```diff
@@ -0,0 +1,316 @@
+    while (inputBuffer.hasRemaining() || (eof && !hadEofDecodeCall)) {
+      CoderResult cr = decoder.decode(inputBuffer, outputBuffer, eof);
+      if (eof) {
+        hadEofDecodeCall = true;
+      }
+
+      if (cr.isMalformed() || cr.isUnmappable()) {
+        reportEncodingProblem();
+
+        if (outputBuffer.remaining() < Encoding_Utils.INVALID_CHARACTER.length()) {
+          growOutputBuffer();
+        }
+        outputBuffer.put(INVALID_CHARACTER);
+        inputBuffer.position(inputBuffer.position() + cr.length());
+      } else if (cr.isUnderflow()) {
+        break;
+      } else if (cr.isOverflow()) {
+        growOutputBuffer();
+      }
+    }
+  }
+
+  /** Returns the amount of bytes that have already been consumed by the decoder. */
+  private int getCurrentInputPosition() {
+    if (inputBuffer == null) return 0;
+    return inputBytesConsumedBeforeCurrentBuffer + inputBuffer.position();
+  }
+
+  private void reportEncodingProblem() {
+    encodingIssuePositions.add(getCurrentInputPosition());
+  }
+
+  /**
+   * Flushes the decoder, growing the buffer as needed to ensure that any additional output from the
+   * decoder fits.
+   */
+  private void flushDecoder() {
+    while (decoder.flush(outputBuffer) == CoderResult.OVERFLOW) {
+      growOutputBuffer();
+    }
+  }
+
+  /**
+   * Ensures that the input buffer has enough free space to hold the number of bytes that we want to
+   * read.
+   *
+   * <p>If necessary, the buffer is allocated or grown, preserving any existing content.
+   *
+   * <p>Assumes that the input buffer is in read mode when the method is called.
+   *
+   * <p>The buffer is in write mode after this call and has enough space to hold {@code bytesToRead}
+   * bytes.
+   */
+  private void ensureInputBufferHasEnoughFreeSpace(int bytesToRead) {
+    if (inputBuffer == null) {
+      inputBuffer = ByteBuffer.allocate(bytesToRead);
+    } else {
+      int freeSpaceInInputBuffer = inputBuffer.capacity() - inputBuffer.remaining();
+
+      // After either compacting the buffer or reallocating it, any remaining input is shifted to
+      // the beginning of the buffer. Thus the bytes that preceded the current position are lost
+      // (because they already have been processed), so we increase the counter to keep the global
+      // position in the input.
+      inputBytesConsumedBeforeCurrentBuffer += inputBuffer.position();
+
+      if (freeSpaceInInputBuffer < bytesToRead) {
+        var old = inputBuffer;
+        inputBuffer = ByteBuffer.allocate(old.remaining() + bytesToRead);
+        inputBuffer.put(old);
+      } else {
+        inputBuffer.compact();
+      }
+    }
+
+    assert inputBuffer.remaining() >= bytesToRead;
+  }
+
+  /**
+   * Increases the capacity of the output buffer, preserving its contents.
+   *
+   * <p>The buffer is assumed to be in write mode when entering this method and is left in write
+   * mode when the method returns.
+   */
+  private void growOutputBuffer() {
```

Comment:
for the sake of not repeating very similar code could we reuse the resize in Encoding_Utils?